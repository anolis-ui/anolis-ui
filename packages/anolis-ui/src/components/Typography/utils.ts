const separateObjValuesMemo = new WeakMap();

// Separate properties with objects (responsive values) as value
export const separateObjValues = (x?: Record<string, any>) => {
  if (x === undefined) return undefined;

  if (separateObjValuesMemo.has(x)) {
    return separateObjValuesMemo.get(x);
  }

  const e = Object.entries(x);

  const a = [];
  const b = [];

  for (let i = 0; i < e.length; i++) {
    if (e[i][1] && typeof e[i][1] === "object") {
      a.push(e[i]);
    } else {
      b.push(e[i]);
    }
  }

  const result = [a, b] as const;
  separateObjValuesMemo.set(x, result);
  return result;
};

const groupByBreakpointMemo = new WeakMap();

// Transform from `width: { sm: 2 }` to `sm: { with: 2 }
export const groupByBreakpoint = (x?: [string, Record<string, any>][]) => {
  if (x === undefined) return undefined;

  if (groupByBreakpointMemo.has(x)) {
    return groupByBreakpointMemo.get(x);
  }

  const result: any = {};

  for (let i = 0; i < x.length; i++) {
    const propKey = x[i][0];
    const keys = Object.keys(x[i][1]);

    for (let ii = 0; ii < keys.length; ii++) {
      const breakpointKey = keys[ii];
      result[breakpointKey] = result[breakpointKey] ?? {};
      result[breakpointKey][propKey] = x[i][1][breakpointKey];
    }
  }

  groupByBreakpointMemo.set(x, result);
  return result;
};

// wrap strings (media queries) in an array with css selector
export const wrapObjsWithSelector = (selector: string, arr: (string | {} | null | undefined)[]) => {
  const result = [];

  for (let i = 0; i < arr.length; i++) {
    if (arr[i] === null || arr[i] === undefined) continue;

    if (typeof arr[i] !== "string") {
      result.push(`${selector} {`, arr[i], "}");
    } else {
      result.push(arr[i]);
    }
  }
  return result;
};
