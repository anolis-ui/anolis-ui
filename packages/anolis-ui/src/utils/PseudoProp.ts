import { PropsOf } from "@emotion/react";
import { x } from "@xstyled/emotion";

export type PseudoProp<C extends keyof JSX.IntrinsicElements = "div", P extends object = {}> = PropsOf<(typeof x)[C]> & P;
