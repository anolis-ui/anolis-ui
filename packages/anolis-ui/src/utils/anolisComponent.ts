import { SystemProps, x } from "@xstyled/emotion";
import { StyledComponent } from "@emotion/styled";
import hoistNonReactStatics from "hoist-non-react-statics";
import { forwardRef, ForwardRefRenderFunction, Ref, RefObject, ReactNode } from "react";
import { SizeVariantProps } from "./theme";

export type AnolisComponent<
  Element extends keyof JSX.IntrinsicElements,
  Props extends object,
  Variant extends string = never,
  Size extends string = never
> =
  StyledComponent<
  SystemProps & SizeVariantProps<Variant, Size> & { as?: React.ElementType } & Props,
  {},
  Omit<JSX.IntrinsicElements[Element], "color">
  >;

export type AnolisComponentProps<
  Element extends keyof JSX.IntrinsicElements,
  Props extends object,
  Variant extends string = never,
  Size extends string = never
> =
  & AnolisBaseProps<Element>
  & SizeVariantProps<Variant, Size>
  & { as?: React.ElementType }
  & Props;

export type AnolisBaseProps<
  E extends keyof JSX.IntrinsicElements,
  V extends string = never,
  S extends string = never
> =
  & SystemProps
  & Omit<JSX.IntrinsicElements[E], "color" | "v" | "s" | "ref">
  & SizeVariantProps<V, S>
  & {
    as?: React.ElementType;
    ref?: Ref<RefOf<E>>;

    children?: ReactNode;
  };

export const anolisComponent = <
  Element extends keyof JSX.IntrinsicElements,
  Props extends AnolisComponentProps<Element, object, string, string>
>(
  tag: Element,
  Component: ForwardRefRenderFunction<RefOf<Element>, Props>
): AnolisComponent<Element, Props, NonNullable<Props["v"]>, NonNullable<Props["s"]>> => {
  const a = forwardRef(hoistNonReactStatics(Component, x[tag]));

  return a as any;
};

export const anolisComp = <T, P = {}>(displayName: string, render: ForwardRefRenderFunction<T, P>) => {
  render.displayName = `anolis(${displayName})`;

  return forwardRef(render);
};

// Way to get from element key strings ("div", "span", "a") their HTML type (HTMLDivElement, HTMLSpanElement, HTMLAnchorElement)
type RefOf<K extends keyof JSX.IntrinsicElements> =
  Exclude<Parameters<Exclude<(JSX.IntrinsicElements[K]["ref"]), undefined | string | null | RefObject<{}>>>[0], null>;
