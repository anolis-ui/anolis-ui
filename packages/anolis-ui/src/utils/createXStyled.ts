/* eslint-disable @typescript-eslint/no-shadow */
import emStyled, { CreateStyled } from "@emotion/styled";
import { css, system } from "@xstyled/emotion";
import { StyleGenerator } from "@xstyled/system";

// xstyled does not expose a way to make custom "styled.*Box". Components made with styled() does not
// support utility props. This file contains simplified functions to create styled() with utility props.
// Original: https://github.com/gregberge/xstyled/blob/main/packages/emotion/src/createStyled.ts

type XCSSFunction = typeof css;

const getCreateStyle = (baseCreateStyle: any, css: XCSSFunction, generator: StyleGenerator) => {
  return (strings: any, ...args: any) => {
    if (Array.isArray(strings)) {
      strings = [...strings, "\n"];
    }
    args = [...args, generator];
    return baseCreateStyle((props: any) =>
      css(strings, ...flattenArgs(args, props))(props)
    );
  };
};

const createShouldForwardProp = (generator: StyleGenerator) => {
  const propSet = new Set(generator.meta.props);
  return (prop: string): boolean =>
    prop !== "as" && !prop.startsWith("$") && !propSet.has(prop);
};

export const createBaseStyled = (css: XCSSFunction, generator: StyleGenerator): CreateStyled => (
  (component: any, options: any) =>
    getCreateStyle(
      emStyled(component, {
        shouldForwardProp: createShouldForwardProp(generator),
        ...options
      }),
      css,
      generator
    )
) as CreateStyled;

export const createXStyled = <TGen extends StyleGenerator>(generator: TGen) => {
  const xstyled = createBaseStyled(css, generator);

  return xstyled;
};

const styleToString = (style: any, props: any): any => {
  if (Array.isArray(style)) {
    return style.reduce((str, style) => str + styleToString(style, props), "");
  }
  if (typeof style === "function") return styleToString(style(props), props);
  return style;
};

const flattenArgs = (arg: any, props: any): any => {
  if (typeof arg === "function") return flattenArgs(arg(props), props);
  if (Array.isArray(arg)) return arg.map((arg) => flattenArgs(arg, props));
  return arg;
};

export const xstyled = createXStyled(system);
