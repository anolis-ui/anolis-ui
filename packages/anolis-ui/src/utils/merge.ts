/**
 * Same as Object.assign but deep merges props with _
 * @deprecated
 */
export const merge = <T>(target: T, ...sources: (T | undefined)[]): T => {
  const next = appendTo({} as T, target);
  sources.forEach(s => appendTo(next, s));
  return appendTo({} as Required<T>, next as Required<T>);
};

const appendTo = <T>(t: T, s: T | undefined) => {
  if (s === undefined) {
    return t;
  }

  _keys(s).reverse().forEach(k => {
    t[k] && k[0] === "_" && typeof t[k] === "object"
      ? merge(t[k], s[k])
      : t[k] = s[k];
  });
  return t;
};

const _keys = <T>(o: T): (keyof T & string)[] =>
  Object.keys(o) as any;

/**
 * Returns a new object with properties of input objects (properties starting with "_" are merged deeply, right-to-left priority).
 */
export const merge2 = <T>(...sources: (T | undefined)[]): T => {
  return _merge2<T>(...[...sources].reverse());
};

export const _merge2 = <T>(...sources: (T | undefined)[]): T => {
  const result: any = {};
  const chain: T[] = [];
  const keys = new Set<string>();

  for (const s of sources) {
    if (s === undefined) continue;
    chain.push(s);
    for (const k of Object.keys(s)) {
      keys.add(k);
    }
  }

  for (const key of keys) {
    // eslint-disable-next-line @typescript-eslint/no-extra-parens
    const shouldDeepMerge = key.startsWith("_") && chain.every((s: any) => s[key] === undefined || (s[key] && typeof s[key] === "object"));

    const value = shouldDeepMerge
      ? _merge2(...chain.map(s => (s as any)[key]))
      : (chain as any[]).find(s => !isNil(s[key]))?.[key];

    !isNil(value) && (result[key] = value);
  }

  return result;
};

const isNil = (x: any): x is undefined | null => x === null || x === undefined;
