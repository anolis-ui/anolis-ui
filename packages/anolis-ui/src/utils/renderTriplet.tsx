import { x } from "@xstyled/emotion";
import { ElementType } from "react";

import renderComponent, { Renderable } from "./renderComponent";
import { ElementProps } from "./theme";

export const renderTriplet = <E extends ElementType>(
  $: E | undefined,
  _: ElementProps<E> | undefined,
  children?: Renderable
) => {
  const Component = typeof $! === "string" ? x[$! as never] : $!;
  return children
    ? <Component {..._!}>{renderComponent(children)}</Component>
    : <Component {..._!} />;
};
