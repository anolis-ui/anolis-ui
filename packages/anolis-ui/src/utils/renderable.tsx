import { ComponentType, ReactNode } from "react";
import { isValidElementType } from "react-is";

export type Renderable = ReactNode | ComponentType;

export const renderable = (C: Renderable): ReactNode =>
  typeof C !== "string" && isValidElementType(C)
    ? <C />
    : (C as ReactNode);
