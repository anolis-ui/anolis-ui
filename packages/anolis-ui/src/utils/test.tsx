import AnolisProvider from "components/AnolisProvider";
import { anolisXstyledTheme } from "defaultTheme";
import { emptyTheme } from "theme";
import { FC, ReactElement, ReactNode } from "react";
import { Queries, queries, render, RenderOptions } from "@testing-library/react";
import { axe, toHaveNoViolations } from "jest-axe";
import renderer from "react-test-renderer";
import { ThemeProvider } from "@xstyled/emotion";

expect.extend(toHaveNoViolations);

const Providers: FC<{ children?: ReactNode }> = ({ children }) => {
  return (
    <ThemeProvider theme={anolisXstyledTheme}>
      <AnolisProvider theme={emptyTheme}>
        {children}
      </AnolisProvider>
    </ThemeProvider>
  );
};

export const customRender = <
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement
>(ui: ReactElement, options?: RenderOptions<Q, Container>) => render(ui, { wrapper: Providers, ...options });

export const testA11y = async (ui: ReactElement) => {
  const { container } = customRender(ui);

  expect(await axe(container)).toHaveNoViolations();
};

export const testSnapshotMatch = (ui: ReactElement) => {
  const tree = renderer.create(
    <ThemeProvider theme={anolisXstyledTheme}>
      <AnolisProvider theme={emptyTheme}>
        {ui}
      </AnolisProvider>
    </ThemeProvider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
};

export * from "@testing-library/react";

export { customRender as render };
