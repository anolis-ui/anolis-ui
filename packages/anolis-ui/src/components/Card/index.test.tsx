import { x } from "@xstyled/emotion";
import { render, testA11y, testSnapshotMatch } from "utils/test";

import { Card } from ".";

test("should render", () => {
  render(
    <Card
      title="Card title"
      media={<x.img src="http://placekitten.com/500/300" alt="cat" />}
      footer={<x.p>Footer</x.p>}
    >
      Text text text
    </Card>
  );
});

test("should conform a11y", async () => {
  await testA11y((
    <>
      <Card
        title="Card title"
        media={<x.img src="http://placekitten.com/500/300" alt="cat" />}
        footer={<x.p>Footer</x.p>}
      >
        Text text text
      </Card>
    </>
  ));
});

test("should match snapshot", () => {
  testSnapshotMatch(
    <Card
      title="Card title"
      media={<x.img src="http://placekitten.com/500/300" alt="cat" />}
      footer={<x.p>Footer</x.p>}
    >
      Text text text
    </Card>
  );
});
