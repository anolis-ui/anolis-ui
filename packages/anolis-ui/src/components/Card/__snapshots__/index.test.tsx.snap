// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: #000f20;
  background: white;
  max-width: 26rem;
  border-radius: 3px;
  --x-shadow: 0 0.25rem 1rem 0 rgba(6,18,39,0.15);
  box-shadow: var(--x-ring-shadow,0 0 #0000),var(--x-shadow);
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
  padding: 2rem;
  padding-bottom: 0;
}

.emotion-2 {
  font-size: 1.25rem;
  line-height: 1.75rem;
  font-weight: 600;
  margin-bottom: -0.75rem;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
}

.emotion-5 {
  padding: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-order: 3;
  -ms-flex-order: 3;
  order: 3;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-order: 4;
  -ms-flex-order: 4;
  order: 4;
  padding: 2rem;
  padding-top: 0;
  font-size: 0.75rem;
  line-height: 1.125rem;
  color: #d5d5e0;
}

<div
  className="emotion-0"
>
  <div
    className="emotion-1"
  >
    <span
      className="emotion-2"
    >
      Card title
    </span>
  </div>
  <div
    className="emotion-3"
  >
    <img
      alt="cat"
      className="emotion-4"
      src="http://placekitten.com/500/300"
    />
  </div>
  <div
    className="emotion-5"
  >
    Text text text
  </div>
  <div
    className="emotion-6"
  >
    <p
      className="emotion-4"
    >
      Footer
    </p>
  </div>
</div>
`;
